File: 01-sorting/go/02-merge-sort.go
package main

import "fmt"

func merge(left []int, right []int) []int {
	i, j := 0, 0
	result := make([]int, 0, len(right)+len(left))
	for i < len(left) && j < len(right) {
		if left[i] <= right[j] {
			result = append(result, left[i])
			i++
		} else {
			result = append(result, right[j])
			j++
		}
	}
	result = append(result, left[i:]...)
	result = append(result, right[j:]...)
	return result
}

func mergeSort(arr []int) []int {
	if len(arr) < 2 {
		return arr
	}
	mid := len(arr) / 2
	left, right := arr[:mid], arr[mid:]
	return merge(mergeSort(left), mergeSort(right))

}

func main() {
	arr := []int{2, 7, 5, 3, 1, 9, 8}
	sorted := mergeSort(arr)
	fmt.Println(sorted)
}

File: 01-sorting/py/02-merge-sort.py
def merge_sort(arr):
    if len(arr) < 2:
        return arr
    middle = len(arr) // 2
    left = merge_sort(arr[:middle])
    right = merge_sort(arr[middle:])
    return merge(left, right)

def merge(left, right):
    i = j = 0
    merged = []
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1
    merged.extend(left[i:])
    merged.extend(right[j:])
    return merged

if __name__ == "__main__":
    arr = [3, 7, 4, 12, 6, 5]
    arr = merge_sort(arr)
    print(arr)